text/0000-simple-postfix-macros.md

```diff
@@ -0,0 +1,171 @@
+- Feature Name: `simple_postfix_macros`
+- Start Date: 2018-05-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow simple postfix macros, of the form `expr.ident!()`, to make macro
+invocations more readable and maintainable in left-to-right method chains.
+
+# Motivation
+[motivation]: #motivation
+
+The transition from `try!` to `?` doesn't just make error handling more
+concise; it allows reading expressions from left to right. An expression like
+`try!(try!(try!(foo()).bar()).baz())` required the reader to go back and forth
+between the left and right sides of the expression, and carefully match
+parentheses. The equivalent `foo()?.bar()?.baz()?` allows reading from left to
+right.
+
+The introduction of `await!` in RFC 2394 brings back the same issue, in the
+form of expressions like `await!(await!(await!(foo()).bar()).baz())`. This RFC
+would allow creating a postfix form of any such macro, simplifying that
+expression into a more readable `foo().await!().bar().await!().baz().await!()`.
+
+Previous discussions of method-like macros have stalled in the process of
+attempting to combine properties of macros (such as unevaluated arguments) with
+properties of methods (such as type-based or trait-based dispatch). This RFC
+proposes a minimal change to the macro system that allows defining a simple
+style of postfix macro, designed specifically for `await!` and for future cases
+like `try!` and `await!`, without blocking potential future extensions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a macro using `macro_rules!`, you can include a first argument
+that uses a designator of `self` (typically `$self:self`). This must appear as
+the first argument, and outside any repetition. If the macro includes such a
+case, then Rust code may invoke that macro using the method-like syntax
+`expr.macro!(args)`. The Rust compiler will expand the macro into code that
+receives the evaluated `expr` as its first argument.
```

Comment:
[Eager expansion? Eager expansion!](https://github.com/rust-lang/rfcs/pull/1628)

Reply:
Or maybe not since the macro doesn't accept tokens in this case, only a single interpolated token representing the whole receiver expression, so it may be okay to expand it later.

`x.identity1!().identity2!()` -> `identity2!( $(x.identity1!()) )` -> `x.identity1!()` -> `identity1!( $(x) )` -> `x`

Reply:
Oh, you actually mean "evaluated", not "expanded" according to the `log_value` example.
I.e. if `$self` is used twice in a macro `m` called like `complex_expr.m!(...)`, then `complex_expr` is *evaluated* only once and temporary variable with its result is automatically created and that variable is passed to the macro and not the expression itself?

That's something seriously novel and untypical for our macro system.
I'd expect the macro author to do it instead.
```rust
macro_rules! log_value {
    ($self:self, $msg:expr) => ({
        let tmp = $self;
        eprintln!("{}:{}: {}: {:?}", file!(), line!(), $msg, tmp);
        tmp
    })
}
```

Reply:
@petrochenkov Normally I'd agree, but I specifically went that route because it avoids any questions about how *much* of the preceding expression gets passed into the macro un-expanded and un-evaluated, which has been at issue in previous discussions of postfix macros.

I think it'd be really *surprising* to write something like `a().b().c().d().m!()` and have `m` receive the whole preceding `a().b().c().d()` (and all arguments to those functions), and potentially evaluate it more than once. That holds doubly true if those preceding bits include macros themselves. If I see `m!(foo())` I expect that `foo()` might not get called; if I see `foo().m!()`, foo seems "outside" the macro, so I think it makes sense to *know* that `foo()` will always get called.

I do agree that the evaluation here makes this different than other macro arguments. However, that's part of how I'm positioning this as "simple" postfix macros; it specifically does *not* try to process the tokens.

Reply:
Ok, if macro expansion generates `let tmp = $self;` automatically (in principle, we can do this because we know that for this syntax both `$self` input and macro output are expressions), then the macro needs to decide how it consumes the result of evaluation - by value, by immutable reference, by mutable reference.

Manually it can be easily done using existing mechanisms:
```rust
let tmp = $self; // consume the receiver
let tmp = &$self; // borrow the receiver
let tmp = &mut $self; // borrow the receiver
```
but if it's done implicitly, then it needs some extra syntax - `$self: &mut self` or something.

Reply:
@petrochenkov My intention was that the macro doesn't get to decide that; `$self` has whatever type the "subject" expression does. If the preceding expression has type `&mut T` then the macro gets a `&mut T`.

If you get a `T` and you *don't* want to consume it, you can always `let tmp = &mut $self` yourself, and then return `$self` at the end after you no longer hold that reference.

The one added complication here: if the subject is mutable, should `$self` be effectively a `let mut`? That would allow, for instance, writing `$self = ...;`. I think the answer is yes, and I think the compiler can do that with the generated internal temporary, but it's worth thinking about.

That said: while I'm proposing doing it this way for simplicity's sake, if you think it's straightforward to do this another way, please let me know and I'm open to the possibility. My main argument against doing so is that doing so seems much harder and I don't want to make the perfect the enemy of the good. If doing so seems *easy* to you, I'd like to take that into account.

Reply:
Please put https://github.com/rust-lang/rfcs/pull/2442/files#r188397176 somewhere in the RFC (perhaps in the alternatives section). I was also super surprised to find out that a mere macro matcher can change evaluation order.

Reply:
@durka I added a rationale for this design choice to the RFC.

Reply:
> If you get a T and you don't want to consume it, you can always let tmp = &mut $self yourself, and then return $self at the end after you no longer hold that reference.

I would like to see an example of code where such a macro has been expanded (or at least, a faithful attempt at representing the generated AST).  I would assume that the macro encloses its output in braces (to allow for a let statement), which means it is forced to move input lvalues.